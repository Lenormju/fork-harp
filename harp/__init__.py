from ._logging import get_logger
from .factory import ProxyFactory

__all__ = [
    ProxyFactory,
    get_logger,
]
