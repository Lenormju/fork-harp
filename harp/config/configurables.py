from typing import Annotated, Generic, Self, Type, TypeVar

from pydantic import BaseModel, Field, field_serializer


class BaseConfigurable(BaseModel):
    @classmethod
    def from_dict(cls: Type[Self], data: dict) -> Self:
        return cls(**data)

    @classmethod
    def from_kwargs(cls: Type[Self], **kwargs) -> Self:
        return cls(**kwargs)


class Configurable(BaseConfigurable):
    class Config:
        extra = "forbid"


TConfigurable = TypeVar("TConfigurable", bound=Configurable)


class Stateful(BaseConfigurable, Generic[TConfigurable]):
    settings: Annotated[TConfigurable, Field(repr=False)]

    @classmethod
    def get_settings_type(cls) -> Type[Configurable]:
        return cls.model_fields["settings"].annotation

    @classmethod
    def from_settings_dict(cls: Type[Self], data: dict) -> Self:
        settings_type = cls.get_settings_type()
        return cls(settings=settings_type(**data))

    @classmethod
    def from_settings_kwargs(cls: Type[Self], **kwargs) -> Self:
        settings_type = cls.get_settings_type()
        return cls(settings=settings_type(**kwargs))

    @field_serializer("settings", when_used="json")
    @classmethod
    def __serialize_settings(cls, settings: TConfigurable):
        BaseType = type(settings).__mro__[type(settings).__mro__.index(Configurable) - 1]
        return BaseType.model_construct(
            **{k: v for k, v in settings.model_dump().items() if k in BaseType.model_fields}
        ).model_dump()
