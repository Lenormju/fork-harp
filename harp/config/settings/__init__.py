from .base import Settings, asdict, settings_dataclass
from .disabled import DisableableBaseSettings, DisabledSettings
from .lazy import Definition, Lazy

__all__ = [
    "Settings",
    "Definition",
    "DisableableBaseSettings",
    "DisabledSettings",
    "Lazy",
    "asdict",
    "settings_dataclass",
]
