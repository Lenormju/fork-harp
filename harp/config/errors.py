from harp.errors import BaseError


class ConfigurationError(BaseError):
    pass


class ConfigurationValueError(ConfigurationError, ValueError):
    pass


class ConfigurationRuntimeError(ConfigurationError, RuntimeError):
    pass


class ConfigurationRemovedSettingError(ConfigurationError):
    pass
