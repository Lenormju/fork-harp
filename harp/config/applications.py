from typing import Type

from rodi import Container
from whistle import IAsyncEventDispatcher

from .events import EVENT_FACTORY_BIND, EVENT_FACTORY_BOUND, EVENT_FACTORY_BUILD, EVENT_FACTORY_DISPOSE
from .settings.base import asdict
from .utils import get_application_type, resolve_application_name


class Application:
    class Settings(dict):
        pass

    class Lifecycle:
        on_bind = None
        """
        Placeholder for factory bind event, happening before the container is built. If set, it will be attached to the
        factory dispatcher automatically.

        """

        on_bound = None
        """
        Placeholder for factory bound event, happening after the container is built. If set, it will be attached to the
        factory dispatcher automatically.
        """

        on_build = None
        """
        Placeholder for factory build event, happening after the kernel is built. If set, it will be attached to the
        factory dispatcher automatically.
        """

        on_dispose = None
        """
        Placeholder for factory dispose event, happening after the kernel is disposed. If set, it will be attached to the
        factory dispatcher automatically, in reverse order of appearance (first loaded application will be disposed last).
        """


class ApplicationsRegistry:
    def __init__(self):
        self._applications = {}

    def __contains__(self, name):
        return name in self._applications

    def __getitem__(self, name):
        return self._applications[name]

    def __iter__(self):
        yield from self._applications

    def __len__(self):
        return len(self._applications)

    def get_full_name(self, name):
        mod_name = self[name].__module__
        if mod_name.endswith(".__app__"):
            return mod_name[:-8]
        return mod_name + "." + self[name].__qualname__

    def add(self, *names):
        for name in names:
            full_name = resolve_application_name(name)
            short_name = full_name.split(".")[-1]

            if short_name not in self._applications:
                self._applications[short_name] = get_application_type(full_name)
            elif self._applications[short_name] != get_application_type(full_name):
                raise ValueError(
                    f"Application {short_name} already registered with a different type ({self._applications[short_name].__module__}.{self._applications[short_name].__qualname__})."
                )

    def remove(self, *names):
        for name in names:
            full_name = resolve_application_name(name)
            short_name = full_name.split(".")[-1]

            if short_name in self._applications:
                del self._applications[short_name]

    def items(self):
        return self._applications.items()

    def keys(self):
        return self._applications.keys()

    def values(self):
        return self._applications.values()

    def lifecycle_for(self, name) -> Type[Application.Lifecycle]:
        return getattr(self._applications[name], "Lifecycle") or Application.Lifecycle

    def settings_for(self, name) -> Type[Application.Settings]:
        return getattr(self._applications[name], "Settings") or dict

    def defaults(self):
        return {name: asdict(self.settings_for(name)()) for name in self._applications}

    def register_events(self, dispatcher: IAsyncEventDispatcher):
        for name, AppType in self.items():
            lifecycle = getattr(AppType, "Lifecycle", None) or Application.Lifecycle

            if on_bind := getattr(lifecycle, "on_bind", None):
                dispatcher.add_listener(EVENT_FACTORY_BIND, on_bind)

            if on_bound := getattr(lifecycle, "on_bound", None):
                dispatcher.add_listener(EVENT_FACTORY_BOUND, on_bound)

            if on_build := getattr(lifecycle, "on_build", None):
                dispatcher.add_listener(EVENT_FACTORY_BUILD, on_build)

        for name, AppType in reversed(self.items()):
            lifecycle = getattr(AppType, "Lifecycle", None) or Application.Lifecycle

            if on_dispose := getattr(lifecycle, "on_dispose", None):
                dispatcher.add_listener(EVENT_FACTORY_DISPOSE, on_dispose)

    def register_services(self, container: Container, config: dict):
        for name, AppType in self.items():
            settings_type = self.settings_for(name)
            local_config = config.get(name, None)
            if settings_type is not dict and local_config:
                container.add_instance(local_config, settings_type)

    def aslist(self):
        return [self.get_full_name(name) for name in self]
