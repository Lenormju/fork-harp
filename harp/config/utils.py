import importlib.util
from typing import TYPE_CHECKING, Type

from .defaults import DEFAULT_NAMESPACES

if TYPE_CHECKING:
    from .applications import Application


def get_application_class_name(name):
    return "".join(map(lambda x: x.title().replace("_", ""), name.rsplit(".", 1)[-1:])) + "Application"


def resolve_application_name(spec):
    if "." not in spec:
        for namespace in DEFAULT_NAMESPACES:
            _candidate = ".".join((namespace, spec))
            if importlib.util.find_spec(_candidate):
                return _candidate
    if importlib.util.find_spec(spec):
        return spec

    raise ModuleNotFoundError(f"No application named {spec}.")


#: Cache for application types
_application_types = {}


def get_application_type(name: str) -> Type["Application"]:
    """
    Returns the application class for the given application name.

    :param name:
    :return:
    """
    if name not in _application_types:
        application_spec = importlib.util.find_spec(name)
        if not application_spec:
            raise ValueError(f'Unable to find application "{name}".')

        try:
            application_module = __import__(".".join((application_spec.name, "__app__")), fromlist=["*"])
        except ModuleNotFoundError as exc:
            raise ModuleNotFoundError(
                f'A python package for application "{name}" was found but it is not a valid harp application. '
                'Did you forget to add an "__app__.py"?'
            ) from exc

        application_class_name = get_application_class_name(name)
        if not hasattr(application_module, application_class_name):
            raise AttributeError(f'Application module for {name} does not contain a "{application_class_name}" class.')

        _application_types[application_spec.name] = getattr(application_module, application_class_name)
        _application_types[application_spec.name].name = application_spec.name

    if name not in _application_types:
        raise RuntimeError(f'Unable to load application "{name}", application class definition not found.')

    return _application_types[name]
