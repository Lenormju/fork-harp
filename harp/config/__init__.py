"""
The Config (:mod:`harp.config`) module provides tools to configure *Applications*.

*Applications* are standard python packages that contains specific definitions to be able to register services and
configure themselves. Every feature of Harp is built as an application, and you will want to write applications if you
want to extend harp's functionality for specific needs.


Contents
--------

"""

__title__ = "Config"

from .applications import Application, ApplicationsRegistry
from .builders import ConfigurationBuilder
from .configurables import Configurable, Service, Stateful
from .events import (
    EVENT_BIND,
    EVENT_BOUND,
    EVENT_READY,
    EVENT_SHUTDOWN,
    OnBindEvent,
    OnBoundEvent,
    OnReadyEvent,
    OnShutdownEvent,
)
from .settings import Definition, DisableableBaseSettings, DisabledSettings, Lazy, Settings, asdict, settings_dataclass
from .utils import get_application

__all__ = [
    "Application",
    "ApplicationsRegistry",
    "Configurable",
    "ConfigurationBuilder",
    "Definition",
    "DisableableBaseSettings",
    "DisabledSettings",
    "EVENT_BIND",
    "EVENT_BOUND",
    "EVENT_READY",
    "EVENT_SHUTDOWN",
    "Lazy",
    "OnBindEvent",
    "OnBoundEvent",
    "OnReadyEvent",
    "OnShutdownEvent",
    "Service",
    "Settings",
    "Stateful",
    "asdict",
    "get_application",
    "settings_dataclass",
]
