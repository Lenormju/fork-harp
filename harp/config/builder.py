import os

from config.common import ConfigurationBuilder as BaseConfigurationBuilder
from config.common import ConfigurationSource


class ConfigurationBuilder(BaseConfigurationBuilder):
    def __init__(self, *sources: ConfigurationSource) -> None:
        super().__init__(*sources)

    def add_examples(self, examples):
        from config.yaml import YAMLFile

        from harp.config import examples as examples_package

        _examples_dirname = os.path.dirname(examples_package.__file__)
        for example in examples or ():
            _example_file = os.path.join(_examples_dirname, f"{example}.yml")
            self.add_source(YAMLFile(_example_file))

    def add_files(self, files):
        for file in files or ():
            _, ext = os.path.splitext(file)
            if ext in (".yaml", ".yml"):
                from config.yaml import YAMLFile

                self.add_source(YAMLFile(file))
            elif ext in (".json",):
                from config.json import JSONFile

                self.add_source(JSONFile(file))
            elif ext in (".ini", ".conf"):
                from config.ini import INIFile

                self.add_source(INIFile(file))
            elif ext in (".toml",):
                from config.toml import TOMLFile

                self.add_source(TOMLFile(file))
            else:
                raise ValueError(f"Unknown file extension: {ext}")

    def add_values(self, values: dict):
        for k, v in values.items():
            self.add_value(k, v)
