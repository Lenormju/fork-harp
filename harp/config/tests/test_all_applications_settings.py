import glob
import inspect
import os.path
from os.path import dirname
from pkgutil import iter_modules

import pytest
from pydantic import ValidationError

from harp import ROOT_DIR
from harp.config import asdict
from harp.utils.config import yaml


def list_submodules(module):
    try:
        modpath = module.__path__
    except AttributeError:
        # not a package, return empty list
        return []

    submodules = []
    for submodule in iter_modules(modpath):
        submodules.append(submodule.name)
    return submodules


all_apps = [
    ".".join(dirname(x).split(os.path.sep)) for x in glob.glob("**/__app__.py", root_dir=ROOT_DIR, recursive=True)
]


def get_all_settings_classes(modname):
    modpath, modattr = modname.rsplit(".", 1)
    module = __import__(modname, fromlist=[modattr])
    for _name, _impl in inspect.getmembers(module, inspect.isclass):
        if not _impl.__module__.startswith(modpath):
            continue
        for submodname in list_submodules(module):
            yield from get_all_settings_classes(f"{modname}.{submodname}")
        yield _name, _impl


# Some settings classes requires some value to have a maening. They are not usually the root settings but some
# subsettings that does not really make sense unless relevant configuration is passed. In that case, we provide the
# minimal set of settings required to instanciate it.
REQUIRED_SETTINGS = {
    "harp_apps.proxy.settings.endpoint.EndpointSettings": {
        "name": "api",
        "port": 4000,
    },
    "harp_apps.proxy.settings.remote.endpoint.RemoteEndpointSettings": {
        "url": "https://www.example.com/",
    },
    "harp_apps.dashboard.settings.auth.User": {
        "password": "secret",
    },
}


@pytest.mark.parametrize("app", all_apps)
def test_all_applications_default_settings(app, snapshot):
    modname = app.rsplit(".", 1)[-1]
    submodules = list_submodules(__import__(app, fromlist=[modname]))
    if "settings" not in submodules:
        pytest.skip(f"No settings submodule found in {app}.")

    all_settings_classes = {}
    for _name, _impl in get_all_settings_classes(app + ".settings"):
        _fullname = f"{_impl.__module__}.{_impl.__qualname__}"
        if _fullname not in all_settings_classes:
            all_settings_classes[_fullname] = _name, _impl

    all_defaults = {}
    for _fullname, (_name, _impl) in all_settings_classes.items():
        if _fullname in REQUIRED_SETTINGS:
            with pytest.raises(ValidationError):
                _impl()
            settings_with_defaults = _impl(**REQUIRED_SETTINGS[_fullname])
        else:
            try:
                settings_with_defaults = _impl()
            except ValidationError as exc:
                raise ValueError(f"Invalid default settings for {_fullname}") from exc
        all_defaults[_fullname] = yaml.dump(asdict(settings_with_defaults))

    assert all_defaults == snapshot
