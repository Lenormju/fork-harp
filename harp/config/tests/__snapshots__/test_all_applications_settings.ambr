# serializer version: 1
# name: test_all_applications_default_settings[harp_apps.acme]
  dict({
    'harp_apps.acme.settings.AcmeSettings': '''
      {}
  
    ''',
  })
# ---
# name: test_all_applications_default_settings[harp_apps.contrib.sentry]
  dict({
    'harp_apps.contrib.sentry.settings.SentrySettings': '''
      dsn: null
  
    ''',
  })
# ---
# name: test_all_applications_default_settings[harp_apps.dashboard]
  dict({
    'harp_apps.dashboard.settings.DashboardSettings': '''
      {}
  
    ''',
    'harp_apps.dashboard.settings.auth.BasicAuthSettings': '''
      type: basic
  
    ''',
    'harp_apps.dashboard.settings.auth.User': '''
      password: secret
  
    ''',
    'harp_apps.dashboard.settings.devserver.DevserverSettings': '''
      {}
  
    ''',
  })
# ---
# name: test_all_applications_default_settings[harp_apps.http_client]
  dict({
    'harp_apps.http_client.settings.HttpClientSettings': '''
      {}
  
    ''',
    'harp_apps.http_client.settings.cache.CacheSettings': '''
      {}
  
    ''',
  })
# ---
# name: test_all_applications_default_settings[harp_apps.janitor]
  dict({
  })
# ---
# name: test_all_applications_default_settings[harp_apps.notifications]
  dict({
    'harp_apps.notifications.settings.NotificationsSettings': '''
      {}
  
    ''',
  })
# ---
# name: test_all_applications_default_settings[harp_apps.proxy]
  dict({
    'harp_apps.proxy.settings.BaseProxySettings': '''
      {}
  
    ''',
    'harp_apps.proxy.settings.Proxy': '''
      settings: {}
  
    ''',
    'harp_apps.proxy.settings.ProxySettings': '''
      {}
  
    ''',
    'harp_apps.proxy.settings.endpoint.BaseEndpointSettings': '''
      name: api
      port: 4000
  
    ''',
    'harp_apps.proxy.settings.endpoint.Endpoint': '''
      settings:
        description: null
        name: api
        port: 4000
  
    ''',
    'harp_apps.proxy.settings.endpoint.EndpointSettings': '''
      name: api
      port: 4000
  
    ''',
    'harp_apps.proxy.settings.remote.BaseRemoteSettings': '''
      {}
  
    ''',
    'harp_apps.proxy.settings.remote.Remote': '''
      current_pool: []
      endpoints: []
      settings:
        break_on:
        - network_error
        - unhandled_exception
        check_after: 10.0
        min_pool_size: 1
  
    ''',
    'harp_apps.proxy.settings.remote.RemoteSettings': '''
      {}
  
    ''',
    'harp_apps.proxy.settings.remote.endpoint.RemoteEndpoint': '''
      settings:
        failure_threshold: 1
        pools:
        - default
        success_threshold: 1
        url: https://www.example.com/
  
    ''',
    'harp_apps.proxy.settings.remote.endpoint.RemoteEndpointSettings': '''
      url: https://www.example.com/
  
    ''',
    'harp_apps.proxy.settings.remote.probe.RemoteProbe': '''
      settings:
        headers: {}
        interval: 10.0
        method: GET
        path: /
        timeout: 10.0
        verify: true
  
    ''',
    'harp_apps.proxy.settings.remote.probe.RemoteProbeSettings': '''
      {}
  
    ''',
  })
# ---
# name: test_all_applications_default_settings[harp_apps.rules]
  dict({
    'harp_apps.rules.models.rulesets.RuleSet': '''
      {}
  
    ''',
    'harp_apps.rules.settings.RulesSettings': '''
      {}
  
    ''',
  })
# ---
# name: test_all_applications_default_settings[harp_apps.storage]
  dict({
    'harp_apps.storage.settings.StorageSettings': '''
      {}
  
    ''',
    'harp_apps.storage.settings.blobs.BlobStorageSettings': '''
      {}
  
    ''',
    'harp_apps.storage.settings.database.DatabaseSettings': '''
      {}
  
    ''',
    'harp_apps.storage.settings.redis.RedisSettings': '''
      {}
  
    ''',
  })
# ---
